components/merkledb/src/views/address.rs

```diff
@@ -0,0 +1,207 @@
+            [prefix, ".", self.name()].concat()
+        };
+        Self {
+            name,
+            id_in_group: self.id_in_group,
+        }
+    }
+
+    /// Appends a name part to `IndexAddress`. The name is separated from the existing name
+    /// by a dot `.`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use exonum_merkledb::IndexAddress;
+    /// let addr = IndexAddress::from_root("foo");
+    /// let suffixed = addr.append_name("suffix");
+    /// assert_eq!(suffixed.name(), "foo.suffix");
+    /// ```
+    pub fn append_name(self, suffix: &str) -> Self {
+        let name = if self.name.is_empty() {
+            suffix.to_owned()
+        } else {
+            // Because `concat` is faster than `format!("...")` in all cases.
+            [self.name(), ".", suffix].concat()
+        };
+        Self {
+            name,
+            id_in_group: self.id_in_group,
+        }
+    }
+
+    /// Appends a key to the `IndexAddress`.
+    pub fn append_key<K: BinaryKey + ?Sized>(self, suffix: &K) -> Self {
+        let name = self.name;
+        let bytes = if let Some(ref bytes) = self.id_in_group {
+            concat_keys!(bytes, suffix)
+        } else {
+            concat_keys!(suffix)
+        };
+
+        Self {
+            name,
+            id_in_group: Some(bytes),
+        }
+    }
+
+    /// Full address with a separator between `name` and `bytes` represented as byte array.
+    pub(crate) fn fully_qualified_name(&self) -> Vec<u8> {
+        /// Separator between the name and the additional bytes in family indexes.
+        const INDEX_NAME_SEPARATOR: &[u8] = &[0];
+
+        if let Some(bytes) = self.id_in_group() {
+            concat_keys!(self.name(), INDEX_NAME_SEPARATOR, bytes)
+        } else {
+            concat_keys!(self.name())
+        }
+    }
+}
+
+impl<'a> From<&'a str> for IndexAddress {
+    fn from(name: &'a str) -> Self {
+        Self::from_root(name)
+    }
+}
+
+impl From<String> for IndexAddress {
+    fn from(name: String) -> Self {
+        Self::from_root(name)
+    }
+}
+
+impl<'a, K: BinaryKey + ?Sized> From<(&'a str, &'a K)> for IndexAddress {
+    fn from((name, key): (&'a str, &'a K)) -> Self {
+        Self {
+            name: name.to_owned(),
+            id_in_group: Some(key_bytes(key)),
+        }
+    }
+}
+
+/// Resolved address of a view.
+///
+/// While an `IndexAddress` is a logical location of a view, a `ResolvedAddress`
+/// represents its location in the key-value storage. Mapping between `IndexAddress`es
+/// and `ResolvedAddress`es is internal to the database logic.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct ResolvedAddress {
+    /// Name of the column family where the view is stored.
+    pub name: String,
+    /// Index identifier within a column family.
+    pub id: Option<NonZeroU64>,
+}
+
+// This conversion is only useful for tests, since all user-created indexes should have an ID set.
+#[cfg(test)]
```

Comment:
Nit/bikeshedding: I'd put `cfg(test)` conversion below `impl ResolvedAddress`, since it's auxiliary.